File: src/api/Dtos/Responses/PagedResult.cs
namespace Example.Api.Dtos.Responses;

/// <summary>
/// Represents the result of a paged query.
/// </summary>
/// <typeparam name="T">The type of items in the result.</typeparam>
public record PagedResult<T>
{
    /// <summary>
    /// Gets the collection of items for the current page.
    /// </summary>
    public IEnumerable<T> Data { get; init; } = [];

    /// <summary>
    /// Gets the current page number.
    /// </summary>
    public int PageNumber { get; init; }

    /// <summary>
    /// Gets the total number of pages.
    /// </summary>
    public int TotalPages { get; init; }

    /// <summary>
    /// Gets the total count of items across all pages.
    /// </summary>
    public long TotalCount { get; init; }

    /// <summary>
    /// Gets the number of items per page.
    /// </summary>
    public int PageSize { get; init; }

    /// <summary>
    /// Gets a value indicating whether there is a previous page.
    /// </summary>
    public bool HasPreviousPage => PageNumber > 1;

    /// <summary>
    /// Gets a value indicating whether there is a next page.
    /// </summary>
    public bool HasNextPage => PageNumber < TotalPages;
}

File: src/nodejs/src/dtos/PagedResult.ts
/**
 * Result DTO for paginated queries.
 * @template T - The type of items in the result.
 */
export class PagedResult<T> {
  /** The items for the current page. */
  data: T[];
  /** The current page number. */
  pageNumber: number;
  /** The total number of pages available. */
  totalPages: number;
  /** The total count of all items across all pages. */
  totalCount: number;
  /** The number of items per page. */
  pageSize: number;

  /**
   * Creates a new paged result.
   * @param data - The items for the current page.
   * @param totalCount - The total count of all items across all pages.
   * @param pageNumber - The current page number.
   * @param pageSize - The number of items per page.
   */
  constructor(data: T[], totalCount: number, pageNumber: number, pageSize: number) {
    this.data = data;
    this.totalCount = totalCount;
    this.pageNumber = pageNumber;
    this.pageSize = pageSize;
    this.totalPages = Math.ceil(totalCount / pageSize);
  }

  /**
   * Indicates whether there is a previous page available.
   */
  get hasPreviousPage(): boolean {
    return this.pageNumber > 1;
  }

  /**
   * Indicates whether there is a next page available.
   */
  get hasNextPage(): boolean {
    return this.pageNumber < this.totalPages;
  }
}
